supportedBackends/tes/src/main/scala/cromwell/backend/impl/tes/TesAsyncBackendJobExecutionActor.scala

```diff
@@ -401,9 +431,41 @@ class TesAsyncBackendJobExecutionActor(override val standardParams: StandardAsyn
 
   override def requestsAbortAndDiesImmediately: Boolean = false
 
+  override def onTaskComplete(runStatus: TesRunStatus, handle: StandardAsyncPendingExecutionHandle): Unit = {
+    val taskEndTime = getTaskEndTime(handle, getTaskLogs)
+    if (runStatus == Error() | runStatus == Failed()) {
+      val errors = getErrorSeq(runStatus, handle, getErrorLogs)
+      errors.onComplete {
+        case Success(r) => tellMetadata(Map(CallMetadataKeys.Failures -> r))
+        case Failure(e) => log.error(e.getMessage)
+      }
+    }
+
+    taskEndTime.onComplete {
+      case Success(result) => tellMetadata(Map(CallMetadataKeys.TaskEndTime -> result))
+      case Failure(e) => log.error(e.getMessage)
+    }
+  }
+
   override def pollStatusAsync(handle: StandardAsyncPendingExecutionHandle): Future[TesRunStatus] =
+    handle.previousState match {
+      case None => pollStatus(handle, fetchCostData = true)
+      case Some(v) =>
+        val validatedCostData = for {
+          costData <- v.costData
+          startTime = costData.startTime
+          vmCost = costData.vmCost
+        } yield (startTime, vmCost)
+
+        validatedCostData match {
+          case Some(_) => pollStatus(handle, fetchCostData = false)
+          case None => pollStatus(handle, fetchCostData = true)
+        }
+    }
+
+  private def pollStatus(handle: StandardAsyncPendingExecutionHandle, fetchCostData: Boolean) =
     for {
-      status <- queryStatusAsync(handle)
+      status <- queryStatusAndCostData(handle, fetchCostData) // Default, minimum data
```

Comment:
This is sort of a nitpick, but I find the name of this method confusing, given that the cost data querying is optional. Cromwell has a lot of names like `queryStatusAndMaybeCostData`.